src/wireviz/wv_dataclasses.py

```diff
@@ -0,0 +1,815 @@
+        else:
+            return self.qty.number
+
+    def bom_amount(self) -> NumberAndUnit:
+        if self.sum_amounts_in_bom:
+            return NumberAndUnit(None, None)
+        else:
+            return self.amount
+
+    @property
+    def has_pn_info(self) -> bool:
+        return any([self.pn, self.manufacturer, self.mpn, self.supplier, self.spn])
+
+
+@dataclass
+class AdditionalComponent(Component):
+    qty_multiplier: Union[QtyMultiplierConnector, QtyMultiplierCable, int] = 1
+    _qty_multiplier_computed: Union[int, float] = 1
+    designators: Optional[str] = None  # used for components definedi in the
+    #                                    additional_bom_items section within another component
+    bgcolor: SingleColor = None  #       ^ same here
+    note: str = None
+
+    def __post_init__(self):
+        super().fill_partnumbers()
+        self.bgcolor = SingleColor(self.bgcolor)
+        self.qty = self.parse_number_and_unit(self.qty, None)
+        self.amount = self.parse_number_and_unit(self.amount, None)
+
+        if isinstance(self.qty_multiplier, float) or isinstance(
+            self.qty_multiplier, int
+        ):
+            pass
+        else:
+            self.qty_multiplier = self.qty_multiplier.upper()
+            if self.qty_multiplier in QtyMultiplierConnector.__members__.keys():
+                self.qty_multiplier = QtyMultiplierConnector[self.qty_multiplier]
+            elif self.qty_multiplier in QtyMultiplierCable.__members__.keys():
+                self.qty_multiplier = QtyMultiplierCable[self.qty_multiplier]
+            else:
+                raise Exception(f"Unknown qty multiplier: {self.qty_multiplier}")
+
+    @property
+    def additional_components(self):
+        # an additional component may not have further nested additional comonents
+        return []
+
+    @property
+    def bom_qty(self):
+        return self.qty.number * self._qty_multiplier_computed
+
+    @property
+    def description(self) -> str:
+        return f"{self.type}{', ' + self.subtype if self.subtype else ''}"
+
+
+@dataclass
+class GraphicalComponent(Component):  # abstract class, for future use
+    bgcolor: Optional[SingleColor] = None
+
+
+@dataclass
+class TopLevelGraphicalComponent(GraphicalComponent):  # abstract class
+    # component properties
+    designator: Designator = None
+    color: Optional[MultiColor] = None
+    image: Optional[Image] = None
+    additional_parameters: Optional[Dict] = None
+    additional_components: List[AdditionalComponent] = field(default_factory=list)
+    notes: Optional[MultilineHypertext] = None
+    # BOM options
+    add_up_in_bom: Optional[bool] = None
+    # rendering options
+    bgcolor_title: Optional[SingleColor] = None
+    show_name: Optional[bool] = None
+
+
+@dataclass
+class Connector(TopLevelGraphicalComponent):
+    # connector-specific properties
+    style: Optional[str] = None
+    category: Optional[str] = None
```

Comment:
Why defining the `category` member here when it's already defined in a superclass?

Reply:
True. Currently, it would even make sense to define it only in `Cable` class (since it is only used there) instead of the `Component` superclass!

Reply:
> True. Currently, it would even make sense to define it only in `Cable` class (since it is only used there) instead of the `Component` superclass!

I agree, but it seems to be used several places outside the `Cable` class. It might be connected to a conflict between two ways of organizing the code:
 - **Object oriented:** 
   * _Encapsulate_ properties and functions that process them into objects with methods.
   * _Abstraction_ - hiding object details that are not needed from the outside.
   * _Inheritance_ between objects when they need a common set of properties and methods.
   * _Polymorphism_ - enable generic algorithms that work on different objects by implementing the code that differ between objects as methods of each object class, and with common method names to be called by the generic algorithms so they don't need to know what kind of object they process.
 - **Procedural programming:**
   * Each procedure/function contains the steps of their algorithm. This is easy to understand as long as the complexity and amount of different alternative value type dependent code don't grow too large.
   * When a lot of the steps in the procedure differ depending on the value type/class being processed, then the code often gets harder to understand due to the needed amount of code to test for different alternatives within each step.

This PR increases the amount of object oriented code by replacing some repeated properties and code with inheritance between classes, but I'm afraid the amount of polymorphism has been reduced, and as a result we see a greater amount of code that test what kind of object is being processed.

Maybe there's a conflict between grouping code together with the algorithm where used, and grouping code together with the object being processed. I feel we should shift this balance a bit in the latter direction in some parts of our code.

Reply:
as a non-professional programmer, thanks for the input and explaining the dualism. I did indeed move in the OO way, but I am not married to the idea and always open for the best approach for the situation.

Reply:
For now, removed the duplicate definition inside `Connector`, will rely on `Component` superclass for now. See https://github.com/wireviz/WireViz/issues/225#issuecomment-951164193 for potential additional uses of the `category` attribute for connectors.